src/soundio/sounddeviceportaudio.cpp

```diff
@@ -430,27 +452,24 @@ SoundDeviceStatus SoundDevicePortAudio::close() {
             return SoundDeviceStatus::Error;
         }
 
-        // Stop the stream. Since the start of this function we are returning
-        // paAbort from the callback. paAbort stops the stream as soon as possible.
-        // When stopping the stream using this technique, we still need to call
-        // Pa_StopStream() before starting the stream again.
+        // Tell the process callback to return paAbort, which will cause
+        // PortAudio to stop the stream as soon as possible. This has the
+        // advantage over calling Pa_StopStream that it will not wait until all
+        // the buffers have been flushed, which can take a few (annoying)
+        // seconds when you're doing soundcard input.
+        m_callbackResult.store(paAbort, std::memory_order_release);
 
+        // We wait until the callback set with Pa_SetStreamFinishedCallback has
+        // been called. Without this, there is the risk of a deadlock when we
+        // call Pa_StopStream and Pa_CloseStream below.
+        waitForFinished();
+
+        // Call Pa_StopStream() in case the above failed.
         err = Pa_StopStream(pStream);
+
         // We should now be able to safely set m_pStream to null
         m_pStream.store(nullptr, std::memory_order_release);
 
-        // Note: With the use of return value paAbort as described above,
-        // the comment below is not relevant anymore, but I am leaving it
-        // because of the BIG FAT WARNING, just in case.
-        //
-        // Trying Pa_AbortStream instead, because StopStream seems to wait
-        // until all the buffers have been flushed, which can take a
-        // few (annoying) seconds when you're doing soundcard input.
-        // (it flushes the input buffer, and then some, or something)
-        // BIG FAT WARNING: Pa_AbortStream() will kill threads while they're
-        // waiting on a mutex, which will leave the mutex in an screwy
-        // state. Don't use it!
-
         if (err != paNoError) {
```

Comment:
This condition is totally unclear for me? What is the reasoning?

Reply:
The result from     resetBlioRingBuffers( &stream->blio );

Reply:
The `if (err != paNoError)` condition? To log it if stop stream fails. But this was already here. Or are you referring to something else?

Reply:
Shouldn't we call  CloseStream() anyway Else we create a memory leak. At least it is called from every error in OpenStream();

Reply:
Ok.